### PR TITLE
pinot plugin loader

### PR DESCRIPTION
**Pull request by rfscholte**
This PR must be merged after #13845 and it assumes that #13849 uses the same structure of the zipfile.

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PinotPluginHandler.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    Path classes = directory.toPath().resolve("classes");
+    if (Files.isDirectory(classes)) {
+      try {
+        urlList.add(classes.toUri().toURL());
+      } catch (MalformedURLException e) {
+        LOGGER.error("Unable to load plugin [{}] classes directory", pluginName, e);
+      }
+    }
+
+    try (var stream = Files.newDirectoryStream(directory.toPath().resolve("lib"), e -> e.getFileName().toString().endsWith(".jar"))) {
```
Comment (gortiz) [inline pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PinotPluginHandler.java:68]: Bad news! We cannot use `var`, given we need to be compatible with Java 11

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginsDirectoryVisitor.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collection;
+import java.util.HashSet;
+
+
+class PluginsDirectoryVisitor extends SimpleFileVisitor<Path> {
```
Comment (gortiz) [inline pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginsDirectoryVisitor.java:12]: nit: I don't like having these individual classes that are designed to resolve a problem that is specific to some other class. It is not important, but I would prefer to refactor this class as a inner class of PluginManager.

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginsDirectoryVisitor.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  private final Collection<Path> _pluginDirs = new HashSet<>();
+
+  private final PinotPluginHandler _pinotPluginHandler;
+
+  private final ShadedPluginHandler _shadedPluginHandler;
+
+  public PluginsDirectoryVisitor(PinotPluginHandler _pinotPluginHandler, ShadedPluginHandler _shadedPluginHandler) {
+    this._pinotPluginHandler = _pinotPluginHandler;
+    this._shadedPluginHandler = _shadedPluginHandler;
+  }
```
Comment (gortiz) [inline pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginsDirectoryVisitor.java:23]: the code style checker is going to complain about that. Attributes must be prefixed with `_` and variables must not be prefixed with `_`. Also `this.attribute` is forbidden.

I hate that code style but... one gets used to it :D

**Comment by rfscholte** [comment]
Superseded by https://github.com/apache/pinot/pull/13930

**Pull request closed without merging**